text/0069-api-v2.md

```diff
@@ -0,0 +1,106 @@
+3. Ensure that no non-TiDB client is online. (Guaranteed by user)
+4. Use `pd-ctl` to enable `API V2`.
+5. Restart all TiKV. (Guaranteed by user)
+
+### Downgrade
+
+Downgrade from `API V2` to `API V1` is also simple:
+
+1. Ensure that no non-TiDB TiKV client is online.
+2. Ensure that no non-TiDB data is in TiKV. Detele if any. (Guaranteed by user)
+3. Use `pd-ctl` to disable `API V2`.
+4. Restart all TiKV. (Guaranteed by user)
+
+### Client
+
+TiKV clients will have a new configuration choosing `API V1` or `API V2` mode on connecting to TiKV.
+
+TiKV will reject the request if the client mode is mismatched to the TiKV configuration. There is one exception: With TiKV in `API V2` mode, you can use `TxnClient` in `API V1` mode to access the TiDB data (key start with `m` or `t`) which is invisible to the client in `API V2` mode.
+
+Note that in `API V2` mode, you can't specify CF in `RawClient`.
+
+Using TxnKV along with RawKV is safe in `API V2` mode.
+
+## Reference-level design
+
+### RawKV Key Encode
+
+Once the `API V2` is enabled, the key will be either start with either:
+
+1. `m` or `t`: TxnKV key. Used by TiDB.
+2. `x\0`: TxnKV key. The `\0` is a placeholder for the potential [Keyspace](https://github.com/tikv/rfcs/pull/39).
+3. `r\0`: RawKV key.
+
+Note that in TxnKV, the key will be encoded by `Memory Comparable Encoding`. But since the `Memory Comparable Encoding` will not change the starting bytes but only add paddings, there won't be any overlap between RawKV and TxnKV.
+
+### RawKV Value Encode
+
+If the key is started with `r`, then the value can be either:
+
+1. `{0x0}{data}`
+2. `{0x1}{TTL expire timestamp}{data}`
+
+### PD
+
+Add a new configuration `enable_api_v2` which can be configured by `pd-ctl`. When switching on or off, PD will synchronize the configuration to all TiKV. PD should scan keys to check that no non-TiDB data exists while switching `enable_api_v2`.
+
+### TiKV
+
+Add a new configuatrion `storage.enable_api_v2`. When enabled, `storage.enable_ttl` must also be enabled. TiKV will carry it's `storage.enable_api_v2` in `PutStore` request. PD will reject the `PutStore` if the switch mismatched.
+
+Add `enable_api_v2` to `SSTMeta` and reject importing SST file if the switch is mismatched.
+
+Add a field `enable_api_v2` to TiKV gRPC's context.
+
+If `stoarge.enable_api_v2` is set:
+
+- If the request's context has `enable_api_v2=false`, reject the request except for it's a TxnKV request and the keys start with `m` or `t`.
+
+- If the request's context has `enable_api_v2=true`, only allow the key to start with `r\0` on RawKV API and only allow the key to start with `t\0` on TxnKV API.
+
+- Run TTL compaction filter only on the keys starting with `r`.
+
+- Use the `API V2` Value encode in `RawStore`, `TTLStore` and `sst_importer`.
+
+If `stoarge.enable_api_v2` is not set:
+
+- Reject all requests with `enable_api_v2` in the context.
+
+### TiKV Clients
```

Comment:
TiKV client needs to consider if it needs to decode keys when it loads a region from PD.

Reply:
I know there is a bug the client may try to decode the start key and end key of a region, which will panic when the boundary is not decodable. But this is not related to the api v2 encode, so I didn't list it here.

Reply:
It's not just a bug. It is a quite more complicated problem. Currently, based on whether TiKV is in txnkv or rawkv mode, the region split keys can be encoded or not encoded. When TiDB loads a region from PD, it should determine if it needs to decode region key based on it uses txnkv or rawkv.

Reply:
I think the fix should change the region cache to not assume the split key is decodable.